File: csharp/1026.maximum-difference-between-node-and-ancestor.cs
/*
 * @lc app=leetcode id=1026 lang=csharp
 *
 * [1026] Maximum Difference Between Node and Ancestor
 */

using System;

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {

    private static int MaxAncestorDiffRecurse(TreeNode curr, int max, int min) {
        if (curr == null) {
            return 0;
        }
        int v = curr.val;
        if (v > max) {
            max = v;
        } else if (v < min) {
            min = v;
        }
        return Math.Max(Math.Abs(max - min), Math.Max(MaxAncestorDiffRecurse(curr.left, max, min), MaxAncestorDiffRecurse(curr.right, max, min)));
    }

    public int MaxAncestorDiff(TreeNode root) {
        return MaxAncestorDiffRecurse(root, root.val, root.val);
    }
}
// @lc code=end

File: java/1026.maximum-difference-between-node-and-ancestor.java
/*
 * @lc app=leetcode id=1026 lang=java
 *
 * [1026] Maximum Difference Between Node and Ancestor
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxAncestorDiff(TreeNode root) {
        return maxAncestorDiff(root, root.val, root.val);
    }

    private static int maxAncestorDiff(TreeNode curr, int max, int min) {
        if (curr == null) {
            return 0;
        }
        int newMax = Math.max(max, curr.val), newMin = Math.min(min, curr.val);
        return Math.max(Math.abs(newMax - newMin), Math.max(maxAncestorDiff(curr.left, newMax, newMin), maxAncestorDiff(curr.right, newMax, newMin)));
    }
}
// @lc code=end

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
